edgehog-device-runtime-forwarder/src/connection.rs

```diff
@@ -68,27 +111,269 @@ impl DerefMut for ConnectionHandle {
+/// respective transport protocol struct.
+#[async_trait]
+pub(crate) trait TransportBuilder {
+    type Connection: Transport;
+
+    async fn build(
+        self,
+        id: &Id,
+        tx_ws: Sender<ProtoMessage>,
+    ) -> Result<Self::Connection, ConnectionError>;
+}
+
+/// For each Connection implementing a given transport protocol (e.g., [`Http`], [`WebSocket`]), it
+/// provides a method returning a [`protocol message`](ProtoMessage) to send to the connections
+/// manager.
+#[async_trait]
+pub(crate) trait Transport {
+    async fn next(&mut self, id: &Id) -> Result<Option<ProtoMessage>, ConnectionError>;
+}
+
+/// Builder for an [`Http`] connection.
+#[derive(Debug)]
+
+pub(crate) struct HttpBuilder {
+    /// To send the request the builder must be consumed, so the option can be replaced with None
+    request: reqwest::RequestBuilder,
+}
+
+impl HttpBuilder {
+    fn new(request: reqwest::RequestBuilder) -> Self {
+        Self { request }
+    }
+}
+
+#[async_trait]
+impl TransportBuilder for HttpBuilder {
+    type Connection = Http;
+
+    /// The id and the channel are only useful for [`WebSocket`] connections.
+    async fn build(
+        self,
+        _id: &Id,
+        _tx_ws: Sender<ProtoMessage>,
+    ) -> Result<Self::Connection, ConnectionError> {
+        Ok(Http::new(self.request))
+    }
+}
+
+impl TryFrom<ProtoHttpRequest> for HttpBuilder {
+    type Error = ConnectionError;
+
+    fn try_from(value: ProtoHttpRequest) -> Result<Self, Self::Error> {
+        value
+            .request_builder()
+            .map(HttpBuilder::new)
+            .map_err(ConnectionError::from)
+    }
+}
+
+/// HTTP connection protocol
+#[derive(Debug)]
+pub(crate) struct Http {
+    /// To send the request the builder must be consumed, so the option can be replaced with None.
+    request: Option<reqwest::RequestBuilder>,
+}
+
+impl Http {
+    /// Store the HTTP request the connection will respond to once executed.
+    pub(crate) fn new(request: reqwest::RequestBuilder) -> Self {
+        Self {
+            request: Some(request),
+        }
+    }
+}
+
+#[async_trait]
+impl Transport for Http {
+    /// Send the [HTTP request](reqwest::Request), wait for a response and return it.
+    #[instrument(skip(self))]
+    async fn next(&mut self, id: &Id) -> Result<Option<ProtoMessage>, ConnectionError> {
+        let Some(request) = self.request.take() else {
+            return Ok(None);
+        };
+
+        trace!("sending HTTP request");
+        let http_res = request.send().await?;
+
+        // create the protobuf response to send to the bridge
+        let proto_res = ProtoHttpResponse::from_reqw_response(http_res).await?;
+
+        debug!("response code {}", proto_res.status());
+
+        let proto_msg = ProtoMessage::Http(ProtoHttp::new(
+            id.clone(),
+            ProtoHttpMessage::Response(proto_res),
+        ));
+
+        Ok(Some(proto_msg))
+    }
+}
```

Comment:
I think you can split this file in 3 different: 

1. connection.rs ( traits, ConnectionHandle.. )
2. http.rs ( Http struct and impl)
3. webscoket.rs ( WebSocket and Impl)
 The right thing would also be to move these file to a separate `connection` module.